radix-engine-stores/src/hash_tree/mod.rs

```diff
@@ -20,55 +23,289 @@ mod test;
+/// within the given `store`.
+/// In a traditional JMT, this inserts a new leaf node for each given "change", together with an
+/// entire new "parent chain" leading from that leaf to a new root node (common for all of them).
+/// In our instantiation of the JMT, we first update all nested per-`ReNodeModule` trees (i.e. of
+/// each {`RENodeId`, `NodeModuleId`} pair encountered in the `changes`), and then we update the
+/// single upper-layer tree (representing all `ReNodeModule`).
+/// All nodes that became stale precisely due to this (i.e. not any previous) operation will be
+/// reported before the function returns (see `WriteableTreeStore::record_stale_node`).
+/// Returns the hash of the newly-created root (i.e. representing state at version
+/// `current_version` + 1).
 ///
 /// # Panics
-/// Panics if a root node for `current_version` does not exist. The caller
-/// should use `None` to denote an empty, initial state of the tree (i.e.
-/// inserting at version 1).
+/// Panics if a root node for `current_version` does not exist. The caller should use `None` to
+/// denote an empty, initial state of the tree (i.e. inserting at version 1).
 pub fn put_at_next_version<S: TreeStore>(
     store: &mut S,
     current_version: Option<Version>,
-    changes: &[(SubstateId, Option<Hash>)],
+    changes: Vec<SubstateHashChange>,
 ) -> Hash {
-    let value_set: Vec<(Hash, Option<(Hash, SubstateId)>)> = changes
-        .iter()
-        .map(|(id, value_hash)| {
-            (
-                hash(scrypto_encode(id).unwrap()),
-                value_hash.map(|value_hash| (value_hash, id.clone())),
-            )
+    let changes_by_re_node_module = index_by_re_node_module(changes);
+    let mut nested_root_changes = Vec::new();
+    for (re_node_module, substate_changes) in changes_by_re_node_module {
+        let nested_root =
+            put_substate_changes(store, current_version, &re_node_module, substate_changes);
+        nested_root_changes.push(IdChange::new(re_node_module, nested_root));
+    }
+    put_changes(
+        store,
+        current_version,
+        current_version.unwrap_or(0) + 1,
+        nested_root_changes
+            .into_iter()
+            .map(|change| to_re_node_change(change))
+            .collect(),
+    )
+}
+
+// only internals below
+
+#[derive(
+    Debug, Clone, Copy, PartialEq, Eq, Hash, ScryptoCategorize, ScryptoEncode, ScryptoDecode,
+)]
+struct ReNodeModule {
+    re_node_id: RENodeId,
+    node_module_id: NodeModuleId,
+}
+
+impl ReNodeModule {
+    fn new(re_node_id: RENodeId, node_module_id: NodeModuleId) -> Self {
+        Self {
+            re_node_id,
+            node_module_id,
+        }
+    }
+}
+
+fn index_by_re_node_module(
+    changes: Vec<SubstateHashChange>,
+) -> HashMap<ReNodeModule, Vec<IdChange<SubstateOffset, Hash>>> {
+    let mut by_re_node_module = HashMap::<ReNodeModule, Vec<IdChange<SubstateOffset, Hash>>>::new();
```

Comment:
We need to avoid HashMap because it creates non-deterministic receipts when iterated over - either we should use IndexMap (more performant, and doesn't require `Ord` on the key) or BTreeMap.

Reply:
I guess given that it's in a private function and we don't expose `HashMap` on the interface, we might be confident it's not used in a way which creates non-determinism, but we agreed it wasn't worth the mental overhead and to simply avoid it going forward 👍

Reply:
> in a private function and we don't expose HashMap on the interface, we might be confident it's not used in a way which creates non-determinism

That gives 0% guarantee on non-det :( (imagine rewriting something from that hashmap into a vec before returning - effectively it shuffles the results)
So I agree, if we are on a war against non-det, then let me switch to `IndexMap` everywhere 👍